src/plugins/ui_actions/public/service/ui_actions_service.ts

```diff
@@ -69,12 +70,12 @@ export class UiActionsService {
     return trigger.contract;
   };
 
-  public readonly registerAction = (action: Action) => {
-    if (this.actions.has(action.id)) {
-      throw new Error(`Action [action.id = ${action.id}] already registered.`);
+  public readonly registerAction = <A extends ActionDefinition<any, any>>(definition: A): void => {
+    if (this.actions.has(definition.id)) {
+      throw new Error(`Action [action.id = ${definition.id}] already registered.`);
     }
 
-    this.actions.set(action.id, action);
+    this.actions.set(definition.id, new ActionInternal(definition));
   };
 
   public readonly attachAction = (triggerId: string, actionId: string): void => {
```

Comment:
hm, actually if you keep the original interface around, we could do something like this:

```ts
  public readonly attachAction = <T extends TriggerId>(
    triggerId: T,
    action: Action<TriggerContextMapping[T]>
  ): void => {
```

And then it would be typesafe!

Reply:
That would be a nice improvement. I see, you've attempted using it in your type improvement follow-up PR, let's add this improvement as part of that PR.